source/game.figure.base.bmx

```diff
@@ -407,68 +405,108 @@ Type TFigureBase extends TSpriteEntity {_exposeToLua="selected"}
 
 
 	Method GetTarget:TFigureTargetBase()
-		if figureTargets.length = 0 then return Null
-		return figureTargets[0]
+		Local target:TFigureTargetBase
+
+		LockMutex(figureTargetsMutex)
+			if figureTargets.length
+				target = figureTargets[0]
+			EndIf
+		UnlockMutex(figureTargetsMutex)
+
+		Return target
 	End Method
 
 
 	Method GetTargetObject:object()
-		if figureTargets.length > 0 and figureTargets[0]
-			return figureTargets[0].targetObj
-		endif
-		return null
+		Local targetObj:Object
+
+		LockMutex(figureTargetsMutex)
+			If figureTargets.length > 0 And figureTargets[0]
+				targetObj = figureTargets[0].targetObj
+			EndIf
+		UnlockMutex(figureTargetsMutex)
+		
+		Return targetObj
 	End Method
 
 
 	'add a target AFTER all others
 	Method AddTarget(target:TFigureTargetBase)
-		figureTargets :+ [target]
+		If Not target Then Return
+		
+		LockMutex(figureTargetsMutex)
+			figureTargets :+ [target]
+		UnlockMutex(figureTargetsMutex)
+	End Method
+
+
+	Method ClearTargets:int()
+		LockMutex(figureTargetsMutex)
+			figureTargets = new TFigureTargetBase[0]
+		UnlockMutex(figureTargetsMutex)
 	End Method
 
 
 	'sets the current target, removes all other targets
 	Method SetTarget(target:TFigureTargetBase)
-		ClearTargets()
-		AddTarget(target)
+		'remove a previous current target first!
+		RemoveCurrentTarget()
+		
+		LockMutex(figureTargetsMutex)
+			If Not target 
+				figureTargets = new TFigureTargetBase[0]
+			Else
+				figureTargets = [target]
+			EndIf
+		UnlockMutex(figureTargetsMutex)
 	End Method
 
 
 	Method PrependTarget(target:TFigureTargetBase)
-		figureTargets = [target] + figureTargets
+		LockMutex(figureTargetsMutex)
+			figureTargets = [target] + figureTargets
+		UnlockMutex(figureTargetsMutex)
 	End Method
 
 
 	Method RemoveCurrentTarget:int()
-		if figureTargets.length = 0 then return False
+		Local result:Int
+		LockMutex(figureTargetsMutex)
+			If figureTargets.length and figureTargets[0] 
+				'inform target
+				figureTargets[0].Abort(self)
+				'actually remove it
+				figureTargets = figureTargets[1..]
+				
+				result = True
+			EndIf
+		UnlockMutex(figureTargetsMutex)
 
-		'inform target
-		if figureTargets[0] then figureTargets[0].Abort(self)
-
-		figureTargets = figureTargets[1..]
-		return True
+		Return result
 	End Method
 
 
 	Method FinishCurrentTarget:int()
-		if figureTargets.length = 0 then return False
+		Local result:Int
+		LockMutex(figureTargetsMutex)
```

Comment:
Wäre es sinnvoll für das Locking (wenn Methodenaufrufe involviert sind, nicht ausschließlich Wertzuweisungen) grundsätzlich ein try-finally-Pattern zu nutzen, um sicherzustellen, dass der Lock in jedem Fall wieder freigegeben wird (selbst wenn ein Aufrufer Exceptionhandling macht)?

Reply:
Die Performanz von Try-Finally ist aber auf jeden Fall etwas schlechter. Bei solchen "seltener" aufgerufenen Funktionen hat die Performanz allerdings keinen so starken Einfluss, man koennte also mit Try-Finally arbeiten.